[6.1.x] SELinux support

a-palchikov opened the pull request:
Add SELinux support to Gravtiy.

The policy itself is available in a [separate repository](github.com/gravitational/selinux/).
Policy is embedded in the binary during the build to make bootstrapping automated (esp. important for additional nodes joined to a cluster).

The most complexity lies mostly in the bootstrapping process which is implemented by the new [selinux](https://github.com/gravitational/gravity/blob/dmitri/6.1.x/3224-selinux-container/lib/system/selinux/bootstrap.go) package.
Due to some peculiarities with using stock SELinux policy tools, the bootstrapping process uses transactions for updates (as individual edit operations will be much slower for the amount of updates Gravity needs). The necessary support (specifically filecontext and port configurations) is available in the [schema](https://github.com/gravitational/gravity/tree/dmitri/6.1.x/3224-selinux-container/lib/system/selinux/internal/schema) package.

### Upgrade
Upgrades will only use SELinux if the cluster has it turned on.

### TODO
Take custom ports (from the application manifest) into account when bootstrapping the local ports.

### Limitations
Currently only redhat/CentOS are supported.

Requires https://github.com/gravitational/planet/pull/542 and https://github.com/gravitational/gravity.e/pull/4273.

a-palchikov committed:
wip: SELinux policy module

a-palchikov committed:
Forward port SELinux changes from 6.0.x

a-palchikov committed:
Refactor policy (use better names, group permissions, introduce more container configuration. Update go-ps and logrus

a-palchikov committed:
Debugging module policy with planet running as container_runtime_t. Relay log file paths to sub-commands executed during install.

a-palchikov committed:
Cam start planet container in the configured domain

a-palchikov committed:
Use systemd-nspawn as a chroot replacement as it can mount locations from host

a-palchikov committed:
Running planet container with confined processes

a-palchikov committed:
Started work to clean up the policy by using macros where necessary or
approriate and removing unnecessary or obsolete rules.
Also starting porting to container-selinux by taking advatanage of
recently introduced domain attribute for Container Engines.

a-palchikov committed:
Update rules that only work with types (and not attributes)

a-palchikov committed:
SElinux policy: more rules for gravity app domain.
Mocve agent user (regular and admin) creation to CreateSite API so it is
consistent and closer to each other.
Mue more log spam in satellite checkers.
Update trace for WrapProxy API.

a-palchikov committed:
Container 'almost' starts - systemd needs to be able to spawn child processes with proper domain

a-palchikov committed:
Planet container is able to start and stop with all services running.
Include 'failed' operation state so that the engine can resume the
operation.

a-palchikov committed:
New SELinux domains for kubernetes components and init-relevant
processes.

a-palchikov committed:
Remove commented out rules. Also remove local checks from CLI installer as it does not play well with resumption.

a-palchikov committed:
Clean up SELinux policy.
Introduce more process domains for planet-specific workloads so these
can be extended w/o touching the stock container-selinux types.

a-palchikov committed:
Label planet-specific locations as gravity_container_file_t.
Fix systemd/runc process labels to avoid unconfined_service_t in the
mix.

a-palchikov committed:
Single-node installations w/o breaks from get-go.

a-palchikov committed:
lib/report: repurpose the syslog export to capture the journal log
unconditionally (to capture as much as possible as opposed for select
SYSLOG_IDENTIFIERs).
install: move preflight chekers to prior to installing the service so
that they are run but not during resumption.
lib/selinux: move SELinux file context/port configuration to a new
package and add a corresponding CLI command.

a-palchikov committed:
Fix handling 'peer auth failed' when joining with an expired token.
wip: adding SELinux bootstrap support.

a-palchikov committed:
SELinux policy: add more rules to transition kubectl symlink to proper
context.
Expand: fail the service if the agent fails to connect to the server due
to permission errors (i.e. expired provisioning token, etc.)

a-palchikov committed:
Move SELinux policy to a dedicated repository.
Start working on 'gravity policy' sub-command.

a-palchikov committed:
* Add generation target for SELinux assets to embed the policy inside
the binary
 * Move default port ranges to lib/schema to be able to generate SELinux
import template
 * Include SELinux static assets

a-palchikov committed:
Exclude SELinux policy generator from the build.
Update make triggers to build when sources change.
Make Operator.ConfigurePackages re-entrant (for install).

a-palchikov committed:
Remove SELinux policy interfaces as these are not essential to install a
policy.
Add SELinux bootstrapping logic to join.

a-palchikov committed:
Build SELinux policy as a gravity package

a-palchikov committed:
Merge with the 6.1.x branch and fix resulted issues

a-palchikov committed:
* Remove unused 'restore-fcontext'.
 * Use a fork of vfsgen to generate byte slices and remove compression.

a-palchikov committed:
Add selinux context to cluster state. Move SELinux assets to per-distribution.

a-palchikov committed:
Revert unnecessary changes. Sync with 6.1.x branch

a-palchikov committed:
Update policy from a branch. Sync SELinux flag in cluster/operation state

a-palchikov committed:
Remove redundant unpack calls to avoid resetting the selinux labels on planet rootfs directory when reinstalling the service

a-palchikov committed:
Adding SELinux support to app hook container. Update site-app to support SELinux.

a-palchikov committed:
Update selinux policy for successful install with docker in selinux mode.

a-palchikov committed:
Make SELinux-specific step in package service reinstall optional and configuration-driven.

a-palchikov committed:
Include directry for logger state and restore file contexts during
bootstrapping phase.

a-palchikov committed:
Remove explicit selinux flag from upgrade command.
Update upgrade to detect selinux support on cluster.
Store selinux support boolean in the cluster record.

a-palchikov committed:
Bind mount a host directory in gravity state directory as kubelet's root-dir

a-palchikov committed:
Update policy

a-palchikov committed:
Make --selinux flag enabled by default (will be used with the implicit
SELinux status check).
Configure root-dir for kubelet to coincide with the default to
workaround an issue with the hardcoded path for device plugin path
(See https://github.com/kubernetes/kubernetes/issues/83226 for
reference).

a-palchikov committed:
WIP: adding custom state directory support in bootstrap script.

a-palchikov committed:
Add support for custom system directory to SELinux bootstrap path

a-palchikov committed:
Update 'package export' to support setting SELinux labels to be able to
deploy agents remotely (expand use-case).
Handle vxlan port configuration properly from the joining agent.

a-palchikov committed:
Fix shrink operation by extracting the gravity binary with a proper SELinux context.

a-palchikov committed:
Update satellite to use the selinux branch rebased on release.
Fix the CLI parser to understand negated boolean flags.
Update selinux policy branch to release.

a-palchikov committed:
Fix tests. Rearrange selinux/* to accommodate Docker build support. Implement the rest of FIXMEs

a-palchikov committed:
Avoid duplicate SELinux bootstrap messages.
Make sure each 'gravity install' or 'gravity join' configures the policy
on first run regardless of the process label.
Expand: only re-run preflight checks if possible.

a-palchikov committed:
Set OS metadata if empty in bootstrap config.
Add missing license headers.

a-palchikov committed:
Remove automatically generated artifacts. Update policy makefile.

a-palchikov commented on build.assets/Makefile:
```diff
@@ -72,7 +75,14 @@ TELEKUBE_TSH_TAG := v3.2.13
 TELEKUBE_PUBLISH_FLAGS ?=
 
+GOCACHE_DOCKER_OPTIONS ?=
+GOCACHE ?= $(HOME)/.cache/go-build
```
This helps tremendously during development and it's off by default (hence should not be an issue on jenkins) so I'm able to opt-in if necessary.

a-palchikov commented on lib/expand/bootstrap.go:
```diff
@@ -200,3 +205,17 @@ func getPeerAddrAndToken(ctx operationContext, role string) (peerAddr, token str
 	return peerAddr, instructions.Token, nil
 }
+
+func patchSELinuxConfig(ctx operationContext) error {
```
Configuring custom values is tricky on joining nodes - in this case, in order to set up the custom `vxlan` port, the installer needs to bootstrap the policy with enough rules to make it to dial the cluster and, after learning about the overrides - update the port configuration with the new value for the port.

r0mant replied:
Could you add a comment about what this function does - "patch selinux config" doesn't tell much.

a-palchikov commented on lib/install/phases/bootstrap.go:
```diff
@@ -197,6 +198,7 @@ func (p *bootstrapExecutor) configureSystemDirectories() error {
 		filepath.Join(stateDir, "planet", "etcd"),
 		filepath.Join(stateDir, "planet", "registry"),
 		filepath.Join(stateDir, "planet", "docker"),
+		filepath.Join(stateDir, "planet", "kubelet"),
```
Needed to have additional directories explicitly created in order to correctly label them (see also for `logrange` below).

a-palchikov commented on lib/rpc/deploy.go:
```diff
@@ -209,9 +210,11 @@ func deployAgentOnNode(ctx context.Context, req DeployAgentsRequest, node, nodeS
 			constants.GravityBin, defaults.SharedExecutableMask,
-			req.GravityPackage, gravityHostPath, defaults.GravityServiceURL).
+			req.GravityPackage, gravityHostPath, defaults.GravityServiceURL,
+			selinux.GravityProcessLabel,
```
Need to have the new binary run with proper SELinux context.

a-palchikov commented on lib/schema/ports.go:
```diff
@@ -0,0 +1,76 @@
+package schema
+
+// DefaultPortRanges defines the list of default ports for the cluster
+var DefaultPortRanges = PortRanges{
```
I had to split ports into groups to support the policy grouping.
Also, certain ports are actually defined in another policy (like `5000` we use for registry - these are grouped in `Reserved`) so it needs to be configurable separately.

a-palchikov commented on lib/system/selinux/bootstrap.go:
```diff
@@ -0,0 +1,525 @@
+`))
+
+// TODO: reflect the actual state of local customizations (incl. custom vxlan port)
+var unloadScript0 = template.Must(template.New("selinux-unload0").Parse(`
```
This is currently under TODO but I'll be returning to using this at some point.

a-palchikov commented on tool/gravity/cli/commands.go:
```diff
@@ -1528,6 +1541,16 @@ type SystemStreamRuntimeJournalCmd struct {
 }
 
+// SystemSelinuxBootstrapCmd configures SELinux file contexts and ports on the node
+type SystemSelinuxBootstrapCmd struct {
```
Since the policy is automatically bootstrapped - there's no way to do this manually with ease. This will allow to generate the local filecontext/port changes script to run along with the policy module.
I might be adding additional command to extract the policy modules as I cannot completely exclude the possibility of something breaking and us needing to load policy manually.

r0mant commented on lib/checks/autofix/autofix.go:
```diff
@@ -45,7 +45,7 @@ func Fix(ctx context.Context, probes []*agentpb.Probe, progress utils.Progress)
 			continue
 		}
-		if err := fixProbe(ctx, probe, progress); err != nil {
+		if err := fixProbe(ctx, probe, progress); err != nil && !trace.IsNotImplemented(err) {
```
So if fixProbe returns "not implemented" error, the probe will be added to fixed probes?

a-palchikov replied:
Hmm - good catch since I wanted to only avoid log spam in the impossible case.

r0mant commented on lib/install/phases/bootstrap.go:
```diff
@@ -351,6 +358,23 @@ func (*bootstrapExecutor) PostCheck(ctx context.Context) error {
 	return nil
 }
 
+func (p *bootstrapExecutor) applySelinuxFilecontexts(ctx context.Context, stateDir string) error {
```
Nit: Selinux -> SELinux, Filecontexts -> FileContexts?

r0mant commented on lib/install/phases/bootstrap.go:
```diff
@@ -351,6 +358,23 @@ func (*bootstrapExecutor) PostCheck(ctx context.Context) error {
+		p.Info("SELinux is disabled.")
+		return nil
+	}
+	out, err := exec.CommandContext(ctx, "restorecon", "-R", "-v", stateDir).CombinedOutput()
```
Could you add a comment here about what "restorecon" command does?

r0mant commented on lib/ops/opsservice/site.go:
```diff
@@ -187,6 +187,10 @@ func (s *site) agentUser() (storage.User, error) {
 	return s.users().GetTelekubeUser(s.agentUserEmail())
 }
 
+func (s *site) selinuxEnabled() (enabled bool) {
```
Nit: seLinux?

r0mant commented on lib/schema/ports.go:
```diff
@@ -0,0 +1,76 @@
+	Installer: []PortRange{
+		{Protocol: "tcp", From: 61009, To: 61010, Description: "installer ports"},
+		{Protocol: "tcp", From: 61022, To: 61025, Description: "installer ports"},
+		{Protocol: "tcp", From: 61080, To: 61080, Description: "installer ports"},
```
Is this actually used anywhere?

a-palchikov replied:
Yes, [it is](https://github.com/gravitational/gravity/blob/dmitri/6.1.x/3224-selinux-container/vendor/github.com/gravitational/teleport/lib/config/configuration.go#L464) - let me know if that's also bogus.

r0mant commented on lib/schema/ports.go:
```diff
@@ -0,0 +1,76 @@
+	},
+	Generic: []PortRange{
+		{Protocol: "tcp", From: 3022, To: 3025, Description: "teleport internal SSH control panel"},
+		{Protocol: "tcp", From: 3080, To: 3080, Description: "teleport Web UI"},
```
I don't think this is used either - Teleport web UI is disabled in Gravity.

a-palchikov replied:
Removed.

r0mant commented on lib/schema/ports.go:
```diff
@@ -0,0 +1,76 @@
+		{Protocol: "tcp", From: 4001, To: 4001, Description: "etcd"},
+		// Defines as afs3_callback_port_t by default
+		{Protocol: "tcp", From: 7001, To: 7001, Description: "etcd"},
+		// Defines as commplex_main_port_t by default
```
Nit: Is "commplex" a typo (for "complex") or not? Also, should it say "Define**d**" here and above?

a-palchikov replied:
No, [it's](https://github.com/fedora-selinux/selinux-policy/blob/rawhide/policy/modules/kernel/corenetwork.te.in#L138) not a typo.

r0mant commented on lib/system/environ/uninstall.go:
```diff
@@ -219,17 +237,20 @@ func dockerInfo() (*utils.DockerInfo, error) {
-	if stateDir, err := state.GravityInstallDir(); err == nil {
-		dirs = append(dirs, stateDir)
+	paths = append(paths, state.StateLocatorPaths...)
+	if utils.Exe.WorkingDir != "/" {
```
What does this test for? Not sure I understand. A comment might be warranted.

r0mant commented on lib/system/selinux/bootstrap.go:
```diff
@@ -0,0 +1,525 @@
+// is supported
+func IsSystemSupported(systemID string) bool {
+	switch systemID {
+	case "centos", "rhel":
```
We probably have constants for these IDs somewhere (where we do OS test).

r0mant commented on lib/system/selinux/bootstrap.go:
```diff
@@ -0,0 +1,525 @@
+}
+
+// Patch executes the patch script with the underlying configuration
+func (r PatchConfig) Patch() error {
```
So this seems to be dealing specifically with custom vxlan ports, so maybe call this function so? PatchVxlan. And update the godoc respectively.

r0mant commented on tool/gravity/cli/clusterupdate.go:
```diff
@@ -392,3 +399,28 @@ Please use the gravity binary from the upgrade installer tarball to execute the
+}
+
+func queryClusterStatus() (status *libstatus.Status, err error) {
+	out, err := exec.Command("gravity", "status", "--output=json").CombinedOutput()
```
Why not use the API? Or localenv.ClusterOperation - seems like you only need the cluster above?

a-palchikov replied:
Not really follow about the `localenv.ClusterOperation` - is this a new API?
The reason this does not use any APIs - the use of an API would require connecting to a port which the current executable might not be capable of - hence, it uses the existing binary (which should be correctly labeled on an installed cluster with SELinux previously enabled) to learn whether it should bootstrap.
If you have a better idea how to work around this chicken and egg problem - I'd be more than grateful.

r0mant commented on tool/gravity/cli/config.go:
```diff
@@ -286,7 +297,9 @@ func (i *InstallConfig) CheckAndSetDefaults() (err error) {
 	if err != nil {
-		return trace.Wrap(err)
+		return trace.Wrap(err, "failed to validate installer directory. "+
+			"Make sure you're running the installer from the directory with the contents "+
```
"Make sure you're launching install from the unpacked cluster image directory." <-- IIRC we wanted to get rid of the "installer" terminology.

Also, do you actually have to launch it from that directory? The install command also accepts a path IIRC.

a-palchikov replied:
It is the default mode of operation - not sure a user would bother specifying the path on a command line although I could hint at that in the message.

r0mant commented on tool/gravity/cli/config.go:
```diff
@@ -436,6 +450,27 @@ func (i *InstallConfig) RunLocalChecks() error {
+		return trace.Wrap(err)
+	}
+	if !selinux.IsSystemSupported(metadata.ID) {
+		log.WithField("os", metadata).Info("No SELinux policy for specified system, will skip.")
```
Will an error be returned if you attempt to launch it with selinux on an unsupported distro? Or it will silently fallback to "no selinux"?

a-palchikov replied:
Since the knob is on automatically, I thought it is wrong to fail with an error - it flips it and continues as if it has been disabled.

r0mant commented on tool/gravity/cli/selinux.go:
```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2019 Gravitational, Inc.
```
2020

r0mant commented on tool/gravity/cli/selinux.go:
```diff
@@ -0,0 +1,96 @@
+			return trace.Wrap(err)
+		}
+	}
+	if procContext["type"] == libselinux.GravityInstallerProcessContext["type"] {
```
What does this check for? A comment would be nice.

r0mant commented on tool/gravity/cli/selinux.go:
```diff
@@ -0,0 +1,96 @@
+	return syscall.Exec(cmd, os.Args, newRespawnEnviron())
+}
+
+func bootstrapSelinux(env *localenv.LocalEnvironment, path, stateDir string, vxlanPort int) error {
```
Nit: SELinux?

r0mant commented on tool/gravity/cli/selinux.go:
```diff
@@ -0,0 +1,96 @@
+}
+
+func newRespawnEnviron() (environ []string) {
+	return append(os.Environ(), fmt.Sprintf("%v=yes", alreadyBootstrappedEnv))
```
I wonder why you decided to use environment variable to pass this info, as opposed to cli flag for example.

a-palchikov replied:
Because it is very simple and local - the command line option would require a lot more changes.

r0mant commented on tool/gravity/cli/status.go:
```diff
@@ -284,6 +285,7 @@ func printStatusText(cluster clusterStatus) {
 }
 
 func printClusterStatus(cluster statusapi.Cluster, w io.Writer) {
+	fmt.Fprintf(w, "SELinux support:\t%v\n", formatSELinuxStatus(cluster.SELinux))
```
Nit: I would only display this if selinux is enabled. Otherwise it's gonna be noise most of the time.

r0mant reviewed (commented):
First pass, gotta look more into lib/selinux package.

a-palchikov committed:
Address review comments. Revisit updates to local SELinux configuration for install/expand/update

a-palchikov committed:
Bump e to include SELinux bootstrap install phase.

Fix SELinux detection and automatically turn it off on distributions
that we do not support SELinux on.
Also, thread SELinux configuration in upgrades properly.

a-palchikov committed:
Avoid restorecon on empty path list

a-palchikov committed:
Expose SELinux state to agent deploying code as it is required to make
proper decision to use additional command line options.
Fix the issue with the joining agent not relaying the fact that SELinux
support is not available to the service.

a-palchikov committed:
Fix directory creation/chowning code to account for new directorys during upgrade

r0mant commented on lib/system/environ/uninstall.go:
```diff
@@ -219,22 +238,31 @@ func dockerInfo() (*utils.DockerInfo, error) {
-	if stateDir, err := state.GravityInstallDir(); err == nil {
-		dirs = append(dirs, stateDir)
+	paths = append(paths, state.StateLocatorPaths...)
+	// do not attempt to remove state directory if started with root
```
Why?

r0mant commented on lib/system/selinux/internal/schema/ports.go:
```diff
@@ -0,0 +1,101 @@
+}
+
+// Parse parses port command from the specified arguments
+func (r *PortCommand) Parse(args []string) error {
```
Doesn't all this rather belong in cli package?

a-palchikov committed:
Sync with upstream

a-palchikov committed:
Revert the ROBOTEST_VERSION knob to be able to manually override it

a-palchikov committed:
Do not name container to avoid conflicts when multiple jenkins jobs are running

a-palchikov committed:
Require service user for system phase during expand

a-palchikov committed:
Use experimental robotest tag with SELinux support

a-palchikov committed:
Bump planet to 6.1.x SELinux branch. Remove --selinux flags from planet as it's still running in server mode so they're not necessary

a-palchikov committed:
Use correct journalctl binary on host

a-palchikov committed:
Fix tests. Move test-only packages to lib/rpc/internal. Update e

a-palchikov committed:
Sync lib/rpc with upstream

a-palchikov committed:
Fix unpack to properly use the accepted tar options

a-palchikov commented:
Closing until we've cleared whether we need to backport this.
